docs/plugins/asg_rollout.md

```diff
@@ -0,0 +1,92 @@
+  def asg_rollout
+    Moonshot::Tools::ASGRollout.new(
+      controller: controller,
+      logical_id: 'APIAutoScalingGroup',
+      # Trigger the worker process to shut down cleanly.
+      pre_detach: -> {
+        instance.exec('systemctl stop my-worker-1')
+      }
+
+      # Ensure that the worker process has completed it's work before terminating.
+      terminate_when: -> {
+        0 == instance.exec('pgrep -f my-worker-1').exitstatus
+      }
+    ).run!
+  end
+end
+
+MyService.start
+```
+
+## Configuration
+
+The ASGRollout object accepts two required keyword arguments:
+  - **controller**: A Moonshot::Controller. If you are using
+    Moonshot::CLI the `controller` method is most often when you want
+    to use.
+  - **logical_id**: The Logical resource ID from the CloudFormation
+    stack to operate on.
+
+By default, the ASGRollout class will build a list of instances with
+non-conforming LaunchConfiguration, then one at a time perform the
+following actions:
+  1. Increase the *Max* and *Desired* sizes of the Auto Scaling Group by one.
+  2. Wait for the new instance to be *InService*.
+  3. Detach a non-conforming instance.
+  4. TODO: Do we need to wait for connection draining or is that
+     handled by the ELB detachment mechiansm?
+  5. Wait the duration of the connecti
+  6. Wait for a new instance to replace it.
+  7. Wait for that new instance to be *InService* in the Auto Scaling Group.
+  8. If the Auto Scaling Group has an associated Elastic Load
+     Balancer, wait for the instance to be *InService* there as well.
+  9. Terminate the detached non-conforming instance.
+  10. If there are other non-conforming instances, go to
+     step 3.
+  11. Restore the *Max* and *Desired* sizes of the Auto Scaling Group
+      to their original values.
+
+A number of optional parameters change the behavior to suit your
+project's needs:
+  - **pre_detach** *(Callable)*: This will be run before step 3
+    above. If it returns `false`, the process will be aborted.
+  - **terminate** *(Callable)*: This will be run to execute step 9. If
+    not specified, the default is to use EC2's Terminate API.
+  - **terminate_when** *(Callable)*: This will be run every 5 seconds
+    prior to step 9. Step 9 will continue as soon as this returns
+    `true`.
+  - **terminate_when_timeout** *(Integer)*: (Default 300s) Number of seconds to
+    wait for any terminate_when lambda to return true. After this
+    timeout, the process will be aborted.
+
+For the lambda / callables above, they are evaluated in a context with
+the following methods available:
+  - **instance** *(TODO)*: An SSH connection to the instance
+    configured using the same logic as Moonshot's SSH feature. (TODO:
+    What class is this? Net::SSH::Thing?)
+  - **ec2** *(Aws::EC2::Client)*: A configured Aws::EC2::Client for the region we're
+    operating in.
+  - **instance_id** *(String)*: The EC2 instance ID.
```

Comment:
Sorry I thought it's an EC2 resource, not a client. Maybe it'd be more convenient to pass an EC2 resource instead of passing a client and an instance ID. It is possible that passing a client is required to perform operations which are only possible via the client. I'm probably too tired to judge this, so I'll just leave it to you.